sinon/lib/stub.py

```diff
@@ -17,118 +17,178 @@ class SinonStub(SinonSpy):
+               The first call has an index of 0.
+
         Return:
-            self (able to be chained)
+            a SinonStub object (able to be chained)
         """
-        self.oncall = n + 1
-        return self
+        cond_oncall = n + 1
+        return SinonStubCondition(copy_of=self, oncall=cond_oncall, cond_args=self._cond_args, cond_kwargs=self._cond_kwargs)
 
-    def onFirstCall(self): #pylint: disable=invalid-name,missing-docstring
-        self.oncall = 1
-        return self
+    def onFirstCall(self): #pylint: disable=invalid-name
+        """
+        Equivalent to stub.onCall(0)
+        """
+        return self.onCall(0)
 
-    def onSecondCall(self): #pylint: disable=invalid-name,missing-docstring
-        self.oncall = 2
+    def onSecondCall(self): #pylint: disable=invalid-name
+        """
+        Equivalent to stub.onCall(1)
+        """
+        return self.onCall(1)
+
+    def onThirdCall(self): #pylint: disable=invalid-name
+        """
+        Equivalent to stub.onCall(2)
+        """
+        return self.onCall(2)
+
+    def returns(self, obj):
+        """
+        Customizes the return values of the stub function. If conditions like withArgs or onCall
+        were specified, then the return value will only be returned when the conditions are met.
+
+        Args: obj (anything)
+        Return: a SinonStub object (able to be chained)
+        """
+        super(SinonStub, self).wrap2stub(lambda *args, **kwargs: obj)
         return self
 
-    def onThirdCall(self): #pylint: disable=invalid-name,missing-docstring
-        self.oncall = 3
+    def throws(self, exception=Exception):
+        """
+        Customizes the stub function to raise an exception. If conditions like withArgs or onCall
+        were specified, then the return value will only be returned when the conditions are met.
+
+        Args: exception (by default=Exception, it could be any customized exception)
+        Return: a SinonStub object (able to be chained)
+        """
+        def exception_function(*args, **kwargs):
+            raise exception
+        super(SinonStub, self).wrap2stub(exception_function)
         return self
 
-    def returns(self, obj):
+class SinonStubCondition(SinonStub):
+    """
+    Allows a new SinonStub object to be created each time the end user specifies a new condition. This is necessary
+    to mimic the behaviour of Sinon.JS.
+
+    Author: Jonathan Benn
+    """
+
+    def __new__(cls, *args, **kwargs):
+        """
+        Override the __new__ provided by SinonBase, since we don't want to do any function wrapping
         """
-        Customizing return of stub function
+        return object.__new__(cls)
 
-        The final chained functions of returns/throws will be triggered
-        If there is some conditions, it will ONLY triggered when meeting conditions
+    def __init__(self, copy_of=None, cond_args=None, cond_kwargs=None, oncall=None):
```

Comment:
we shouldn't put `copy_of=None` here, because copy_of couldn't be `None` or the whole class will be broken unexpectedly.

Reply:
You are right. In fact, all of the arguments are not optional. Thanks